### PR TITLE
add builtin-reverse func

### PR DESCRIPTION
**Pull request by mrmiywj**
#310

### CODE REVIEW

File: evaluator/builtin_string_test.go [hunk 1]
```diff
@@ -210,6 +210,29 @@ func (s *testEvaluatorSuite) TestLowerAndUpper(c *C) {
...
+	d, err := builtinReverse(types.MakeDatums([]interface{}{nil}...), nil)
+	c.Assert(err, IsNil)
+	c.Assert(d.Kind(), Equals, types.KindNull)
+
+	tbl := []struct {
+		Input  interface{}
+		Expect string
+	}{
+		{"abc", "cba"},
+		{"LIKE", "EKIL"},
```
Comment (shenli) [inline evaluator/builtin_string_test.go]: Please add more cases such as reverse(null), reverse(123).

File: parser/parser.y [hunk 2]
```diff
@@ -248,6 +248,7 @@ import (
 	rowFormat	"ROW_FORMAT"
 	rsh		">>"
 	rtrim		"RTRIM"
+  reverse "REVERSE"
```
Comment (shenli) [inline parser/parser.y]: Keep alignment.

File: parser/parser.y [hunk 3]
```diff
@@ -2561,6 +2562,10 @@ FunctionCallNonKeyword:
 		args := []ast.ExprNode{$3.(ast.ExprNode), $5.(ast.ExprNode), $7.(ast.ExprNode)}
 		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1.(string)), Args: args}
 	}
+| "REVERSE" '(' Expression ')'
```
Comment (shenli) [inline parser/parser.y]: Keep alignment.

**Comment by shenli** [comment]
Please add type inference here: https://github.com/pingcap/tidb/blob/master/optimizer/typeinferer.go#L281 for reverse().  Meanwhile, add test case here: https://github.com/pingcap/tidb/blob/master/optimizer/typeinferer_test.go#L35

File: parser/parser.y (hunk 3 above)
Comment (shenli) [inline parser/parser.y]: reverse is an unreserved keyword. So please add it into https://github.com/pingcap/tidb/blob/master/parser/parser.y#L1908. Meanwhile add a test case here: https://github.com/pingcap/tidb/blob/master/parser/parser_test.go#L37

File: Makefile [hunk 4]
```diff
@@ -56,6 +56,7 @@ TEMP_FILE = temp_parser_file
 parser:
 	go get github.com/qiuyesuifeng/goyacc
 	go get github.com/qiuyesuifeng/golex
+	go get github.com/golang/example/stringutil
```
Comment (shenli) [inline Makefile]: We use vendor to solve dependency. Should we add it into .vendor? @coocood

**Comment by shenli** [comment]
@mrmiywj Thanks for your contribute!

@coocood @zimulala @zxylvlp PTAL

**Comment by coocood** [comment]
@mrmiywj 
Can you implement the simple function in stringutil directly?
So we don't have to import the package.

File: Makefile (hunk 4 above)
Comment (coocood) [inline Makefile]: It's an example package, it's not supposed to be imported.

**Comment by mrmiywj** [comment]
cc @shenli @coocood

**Comment by shenli** [comment]
LGTM

File: parser/parser.y [hunk 5]
```diff
@@ -2561,6 +2562,10 @@ FunctionCallNonKeyword:
 		args := []ast.ExprNode{$3.(ast.ExprNode), $5.(ast.ExprNode), $7.(ast.ExprNode)}
 		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1.(string)), Args: args}
 	}
+|       "REVERSE" '(' Expression ')'
+        {
+                $$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1.(string)), Args: []ast.ExprNode{$3.(ast.ExprNode)}}
+        }
```
Comment (coocood) [inline parser/parser.y]: In parser.y and scanner.y, we use tab to align the code.
Please change spaces to tabs.

**Comment by mrmiywj** [comment]
cc @coocood 
Sorry for that on my emacs configuration, tabs are automatically replaced by spaces

**Comment by shenli** [comment]
PTAL @coocood

File: parser/scanner.l [hunk 6]
```diff
@@ -445,6 +445,7 @@ references	{r}{e}{f}{e}{r}{e}{n}{c}{e}{s}
 regexp		{r}{e}{g}{e}{x}{p}
 replace		{r}{e}{p}{l}{a}{c}{e}
 redundant	{r}{e}{d}{u}{n}{d}{a}{n}{t}
+reverse {r}{e}{v}{e}{r}{s}{e}
```
Comment (shenli) [inline parser/scanner.l]: Seems it is not aligned.

**Comment by mrmiywj** [comment]
cc @shenli

**Comment by coocood** [comment]
@mrmiywj 
Still use spaces, not aligned.

**Comment by mrmiywj** [comment]
cc @coocood

**Comment by coocood** [comment]
@mrmiywj 
`
{reverse}   lval.item = string(l.val)
      return reverse
`
in `scanner.l`

**Comment by coocood** [comment]
also missed a tab before
`	$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1.(string)), Args: []ast.ExprNode{$3.(ast.ExprNode)}}`

**Comment by mrmiywj** [comment]
cc @coocood

**Comment by coocood** [comment]
@mrmiywj 
The alignment is ok now.
Please resolve the conflict.

**Comment by mrmiywj** [comment]
cc @coocood

**Comment by coocood** [comment]
LGTM
Thank you for your contribution!

**Pull request merged**